### ISSUE TITLE
Form->control on json (array) entity fields

### ISSUE THREAD

**Issue by kolorafa** [cakephp/cakephp#18055]
### Description

Related to #17886 

## What I have

Table with `jsonTypeField` field of type `json` that is automatically converted to array on read with `ORM.mapJsonTypeForSqlite`.

```php  
    echo $this->Form->control('jsonTypeField');
```
### What happens

Warning (2) : Array to string conversion [in /app/vendor/cakephp/cakephp/src/View/StringTemplate.php, line 326]

## What I expected to happen

Good question, probably convert array back to json string to make it still work as plain text (and set the resulting json as input text with some additional marking?)

In one way it's nice that it is json_decode'd when read, it make it easier to access specific fields in code.
But sometimes there is a need to just present raw json to user, as this field is not handled by CakePHP directly (other than doing a json validation).

The whole behavior also relates to #18054 

### CakePHP Version

5.1.2

### PHP Version

8.3.14

**Comment by markstory** [comment]
Moving to enhancement as this isn't a current feature of the framework. Generating inputs for arbitrarly shaped JSON is not within the current FormHelper feature set.

> But sometimes there is a need to just present raw json to user, as this field is not handled by CakePHP directly (other than doing a json validation).

You could handle this with `$this->Form->control('jsonTypeField, ['value' => json_enode($entity->json_field_type)]);`.

**Comment by kolorafa** [comment]
> You could handle this with $this->Form->control('jsonTypeField, ['value' => json_enode($entity->json_field_type)]);.

Yes, that's correct, but in my case tables that I handle are dynamic - so I don't create views for them, so we are looking of making them more "magic/automatic", we already accomplished some of it, but was thinking of polishing it and upstreaming it :)

> Generating inputs for arbitrarly shaped JSON is not within the current FormHelper feature set.

Generating multiple inputs would be a nightmare, I was more thinking of detecting array (with only scalar values) and generating single textarea that can be easily handled by jsoneditor and such. Would the project be interested in such feature?

**Comment by markstory** [comment]
> Generating multiple inputs would be a nightmare, I was more thinking of detecting array (with only scalar values) and generating single textarea that can be easily handled by jsoneditor and such. Would the project be interested in such feature?

That could be good as a new input widget type that developers can enable with something like

```php
$this->Form->control('jsonTypeField', ['type' => 'jsonTextarea'])
```

**Comment by kolorafa** [comment]
My full solution (that automatically set json textarea when database column type is json) is:

Created `APP/src/View/Widget/JsonWidget.php`:

```
use Cake\View\Form\ContextInterface;
use Cake\View\Widget\TextareaWidget;

class JsonWidget extends TextareaWidget
{
    /**
     * @param array $data
     * @param \Cake\View\Form\ContextInterface $context
     * @return string
     */
    public function render(array $data, ContextInterface $context): string
    {
        if (is_array($data['val'])) {
            $data['val'] = json_encode($data['val']);
        }

        return parent::render($data, $context);
    }
}
```

in `APP/src/View/View.php` added:

```
'widgets' => [
   'json' => ['Json'],
],
```

in APP/src/View/Helper/FormHelper.php:
```
        $config['typeMap']['json'] = 'json';
```


in app/config/bootstrap.php:
TypeFactory::map('json', \namespace\JsonType::class);


created `APP/src/ORM/Types/JsonType.php`

```
class JsonType extends JsonType
{
    /**
     * Marshals request data into a JSON compatible structure.
     *
     * @param mixed $value The value to convert.
     * @return mixed Converted value.
     */
    public function marshal(mixed $value): mixed
    {
        if (is_string($value) && json_validate($value)) {
            return json_decode($value, true);
        }

        return parent::marshal($value);
    }
}
```

**Comment by dereuromark** [comment]
You seem to have found a working solution. Nothing else needed it seems.
